components/lark-error/src/lib.rs

```diff
@@ -35,31 +35,45 @@ use std::sync::Arc;
 /// Unit type used in `Result` to indicate a value derived from other
 /// value where an error was already reported.
 #[derive(Clone, Debug, DebugWith, PartialEq, Eq, Hash, PartialOrd, Ord)]
-pub struct ErrorReported(pub Vec<Span>);
+pub struct ErrorReported(pub Vec<LabeledSpan>);
 
 impl ErrorReported {
-    pub fn at_span(s: Span) -> Self {
+    pub fn at_labeled_span(s: LabeledSpan) -> Self {
         ErrorReported(vec![s])
     }
 
-    pub fn at_spans(s: Vec<Span>) -> Self {
+    pub fn at_labeled_spans(s: Vec<LabeledSpan>) -> Self {
         ErrorReported(s)
     }
 
-    pub fn some_span(&self) -> Span {
+    pub fn some_labeled_span(&self) -> LabeledSpan {
         // Pick the first error arbitrarily
-        self.spans()[0]
+        self.labeled_spans()[0].clone()
     }
 
-    pub fn spans(&self) -> &[Span] {
+    pub fn labeled_spans(&self) -> &[LabeledSpan] {
         &self.0
     }
 
-    pub fn into_spans(self) -> Vec<Span> {
+    pub fn into_labeled_spans(self) -> Vec<LabeledSpan> {
         self.0
     }
 }
 
+/// A span with an associated label.
+/// TODO: We may want to merge this with what's available in error
+/// reporting
+#[derive(Clone, Debug, DebugWith, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub struct LabeledSpan {
```

Comment:
Nit: Maybe we can name this "Diagnostic"? Feels like a "nicer word" to me.

Reply:
Sure, can do. I'm guessing it'll grow into that at some point, so earlier is fine with me